Design/indexing_rollout_notes.md

```diff
@@ -0,0 +1,209 @@
+  IndexInfo group(const IndexInfo &, const std::vector<SpectrumIndexSet> &);
+  }
+  ```
+
+- Should we set default detector IDs?
+  This is was workspaces are doing currently, but it does not really make sense.
+  Probably we should find a better way if setting sensible detector IDs once an instrument has been set.
+  This is something that potentially needs to be decided together with other decisions regarding Instrument-2.0.
+
+
+## NearestNeighbours
+
+- Uses `specnum_t` as map key.
+- Should simply use spectrum index?
+- Would that make trouble with MPI? Need to have a detailed look.
+- See issue [#17372](https://github.com/mantidproject/mantid/issues/17372).
+
+
+## Logging and exception messages
+
+- Frequently spectrum numbers are included in log messages or exceptions.
+
+- Probably the easiest (overall) solution is to add means of getting this information automatically, e.g., with a helper function
+
+  ```cpp
+  namespace Logging {
+  std::string spectrumNumber(const MatrixWorkspace &workspace, size_t index);
+  }
+  ```
+
+## File- and UI output
+
+Similar to logging, there are also several algorithms that write, e.g., detector IDs to a file.
+Maybe a similar or even the same helper function as suggested above in the logging section could be used.
+
+
+## Obtaining detector information
+
+Frequently, detector IDs are used to obtain detector pointers or other detector-related information from the instrument.
+The IDs are not actually required here.
+Refactoring would be significantly easier (and would avoid the need to revisit and incrementally refactor) if such information was available via, e.g., `MatrixWorkspace::SpectrumInfo`.
+
+Commonly used functionality includes:
+
+```cpp
+class SpectrumInfo {
+  bool isMasked(const size_t index) const;
+  bool isMonitor(const size_t index) const;
+  double l2(const size_t index) const;
+  Detector_sptr uniqueDetector(const size_t index) const; // Throws if no or more than 1 detector
+  bool hasUniqueDetector(const size_t index) const; // Should this be part of IndexInfo instead?
+};
+```
+
+There is also another (special?) case (example from current code):
+
+```cpp
+Geometry::IDetector_const_sptr det;
+size_t numDetectors = workspace->getSpectrum(wi).getDetectorIDs().size();
+if (numDetectors > 1) {
+  Instrument_const_sptr inst = workspace->getInstrument();
+  det = inst->getDetector(*workspace->getSpectrum(wi).getDetectorIDs().begin());
+} else {
+  det = workspace->getDetector(wi);
+}
+```
+
+## Duplicate spectrum numbers
+
+Currently quite a few algorithms or tests use or create workspaces with non-unique spectrum numbers.
+Ultimately we want to forbid this, but it is not yet clear if we should attempt to clean this up right away, or wait until `IndexInfo` has developed into a more stable state.
+
+
+## Standards
+
+#### Spectrum numbers starting at 0
+
+Currently quite a few algorithms that create workspaces where spectrum numbers do not start at 1 but at 0.
```

Comment:
Well, there are algorithms that create workspaces without user input such as files. In particular, the example I gave (`AppendSpectra`) merged spectra from workspaces, but for some reason the spectrum numbers start at zero. That is, spectrum numbers actually are used directly in this algorithm, since it has to set them to a non-default value. If we would use defaults, we could probably eliminate most of such code.